libraries/chain/block_log.cpp

```diff
@@ -477,22 +483,256 @@ namespace eosio { namespace chain {
                                                                                      uint64_t first_block_position) {
       return reverse_block_position_iterator<BlockLogData>(t, first_block_position);
    }
+
+   template <typename Lambda>
+   void for_each_file_in_dir_matches(const fc::path& dir, const char* pattern, Lambda&& lambda) {
+      const std::regex my_filter(pattern);
+      std::smatch      what;
+      boost::filesystem::directory_iterator end_itr; // Default ctor yields past-the-end
+      for (boost::filesystem::directory_iterator p(dir); p != end_itr; ++p) {
+         // Skip if not a file
+         if (!boost::filesystem::is_regular_file(p->status()))
+            continue;
+         // skip if it's not match blocks-*-*.log
+         if (!std::regex_match(p->path().filename().string(), what, my_filter))
+            continue;
+         lambda(p->path());
+      }
+   } 
    } // namespace
 
-   block_log::block_log(const fc::path& data_dir)
-       : my(new detail::block_log_impl(data_dir)) {
-   }
+   struct block_log_catalog {
+      using block_num_t = uint32_t;
+
+      struct mapped_type {
+         block_num_t             last_block_num;
+         boost::filesystem::path filename_base;
+      };
+      using collection_t              = boost::container::flat_map<block_num_t, mapped_type>;
+      using size_type                 = collection_t::size_type;
+      static constexpr size_type npos = std::numeric_limits<size_type>::max();
+
+      boost::filesystem::path    archive_dir;
+      size_type                  max_retained_files = 10;
+      collection_t               collection;
+      size_type                  active_index = npos;
+      block_log_data             log_data;
+      block_log_index            log_index;
+      chain_id_type              chain_id;
+
+      bool empty() const { return collection.empty();  }
+
+      void open(fc::path block_dir) {
+         for_each_file_in_dir_matches(block_dir, R"(blocks-\d+-\d+\.log)", [this](boost::filesystem::path path) {
+            auto log_path               = path;
+            auto index_path             = path.replace_extension("index");
+            auto path_without_extension = log_path.parent_path() / log_path.stem().string();
+
+            block_log_data log(log_path);
+
+            if (chain_id.empty()) {
+               chain_id = log.chain_id();
+            } else {
+               EOS_ASSERT(chain_id == log.chain_id(), block_log_exception, "block log file ${path} has a different chain id",
+                           ("path", log_path.generic_string()));
+            }
+
+            // check if index file matches the log file
+            if (!index_matches_data(index_path, log))
+               block_log::construct_index(log_path, index_path);
+
+            auto [itr, _]      = collection.emplace(log.first_block_num(),
```

Comment:
Need to check if iterator is not added and if it was not, then need to log a warning and set active_index to npos, since the cached information isn't really in the collection.